eclipse-che/che-dashboard
# feat: set up Git remotes from factory url

Signed-off-by: dkwon17 <dakwon@redhat.com>



### What does this PR do?
Adds a `remotes` url parameters for the factory url to set up remotes in the devworkspace (in the `spec.template.projects` section)

**IMPORTANT! A few notes about this PR:**
* If the repo has no devfile, the default devfile is used and  a project will be added with the remotes configured
* If the repo's devfile has no `projects` field, a project will be added with the remotes configured
* If the repo's devfile has a `projects` field with one `git` project, that project will have the remotes configured
* If the repo's devfile has a `projects` field with more than one `git` project, remotes are not configured since it is unclear what project to configure remotes for.
  * [error message](https://user-images.githubusercontent.com/83611742/208253900-a590a655-2ebf-43af-b9e3-67c3f0908cb2.png)
  * If this is a problem, we can change the syntax used for `remotes` url parameter to specify a project name (`spec.template.projects[*].name`) for the remotes to specifically configure

* It is not yet possible to set what exact branch to checkout, as this PR only introduces the ability to configure remotes. The default branch for the first remote will be checked out when starting the workspace.

### What issues does this PR fix or reference?
Part of https://issues.redhat.com/browse/CRW-3482 and https://github.com/eclipse/che/issues/21315

### Is it tested? How?

As explained in https://issues.redhat.com/browse/CRW-3482, remotes can be configured in the url as follows:
```
# 1. explicit name and URL
?remotes={{origin,https://github...},{upstream,https://github...},...}

# 2. URLs only, name is implicit: first is origin, second is upstream, subsequent are fork1, fork2 etc... 
?remotes={https://github...,https://github...,...}
```

Deploy Che and set the following in the CheCluster CR:
```
spec:
  components:
    dashboard:
      deployment:
        containers:
          - image: 'quay.io/dkwon17/che-dashboard:remotes'
            name: che-dashboard
```

To test, are some scenarios:

1. URLs only, name is implicit.
  Go to: `{CHE-HOST}/#https://github.com/eclipse-che/che-plugin-registry?remotes={https://github.com/dkwon17/che-plugin-registry,https://github.com/eclipse-che/che-plugin-registry,https://github.com/che-dockerfiles/che-plugin-registry}`
  And verify that the resulting devworkspace's yaml has the following remotes:
```
    projects:
      - git:
          checkoutFrom:
            remote: origin
          remotes:
            fork1: 'https://github.com/che-dockerfiles/che-plugin-registry'
            origin: 'https://github.com/dkwon17/che-plugin-registry'
            upstream: 'https://github.com/eclipse-che/che-plugin-registry'
        name: che-plugin-registry
```
2. Explicit name and URL.
  Go to: `{CHE-HOST}/#https://github.com/eclipse-che/che-plugin-registry?remotes={{origin,https://github.com/dkwon17/che-plugin-registry},{upstream,https://github.com/eclipse-che/che-plugin-registry}}`
  And verify that the resulting devworkspace's yaml has the following remotes:
```
    projects:
      - git:
          checkoutFrom:
            remote: origin
          remotes:
            origin: 'https://github.com/dkwon17/che-plugin-registry'
            upstream: 'https://github.com/eclipse-che/che-plugin-registry'
```
3. Explicit name and URL, but change the order of the remotes to verify that `checkoutFrom` has changed.
  Go to: `{CHE-HOST}/#https://github.com/eclipse-che/che-plugin-registry?remotes={{upstream,https://github.com/eclipse-che/che-plugin-registry},{origin,https://github.com/dkwon17/che-plugin-registry}}`
```
    projects:
      - git:
          checkoutFrom:
            remote: upstream
          remotes:
            origin: 'https://github.com/dkwon17/che-plugin-registry'
            upstream: 'https://github.com/eclipse-che/che-plugin-registry'
```
4. If the remotes cannot be parsed due to syntax errors, or if it is empty, this error appears:
![image](https://user-images.githubusercontent.com/83611742/207752368-a32e033f-d5ab-41b4-83c3-868a3741f0fa.png)

#### Release Notes



#### Docs PR
